### PR TITLE
server: ignore raft messages if member id mismatch

### LINKED ISSUE #16220: tocommit(3730) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost
### What would you like to be added?

The heartbeat sent by the leader node contains the committed log index. If the index is higher than the follower node, follower will panic.
ETCD should recheck the logs instead of exiting process.

### Why is this needed?

We deployed a 3-node cluster and tested the cluster. One of the test cases is 65% of the network error packets between nodes. About 10 minutes later, a follower node restarted. The log prints the panic stack:

`msg:
tocommit(3730) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?
stacktrace:
vendor/go.etcd.io/etcd/server/v3/etcdserver.(*zapRaftLogger).Panicf
vendor/go.etcd.io/etcd/server/v3/etcdserver/zap_raft.go:101

vendor/go.etcd.io/etcd/raft/v3.(*raftLog).commitTo
vendor/go.etcd.io/etcd/raft/v3/log.go:237

vendor/go.etcd.io/etcd/raft/v3.(*raft).handleHeartbeat
vendor/go.etcd.io/etcd/raft/v3/raft.go:1509

vendor/go.etcd.io/etcd/raft/v3.stepFollower
vendor/go.etcd.io/etcd/raft/v3/raft.go:1435

vendor/go.etcd.io/etcd/raft/v3.(*raft).Step
vendor/go.etcd.io/etcd/raft/v3/raft.go:975

vendor/go.etcd.io/etcd/raft/v3.(*node).run
vendor/go.etcd.io/etcd/raft/v3/node.go:356`

We have found similar issue #13509, #15699

### LINKED ISSUE #17081: Wrong raft messages may cause etcd panic
### Bug report criteria

- [X] This bug report is not security related, security issues should be disclosed privately via security@etcd.io.
- [X] This is not a support request or question, support requests or questions should be raised in the etcd [discussion forums](https://github.com/etcd-io/etcd/discussions).
- [X] You have read the etcd [bug reporting guidelines](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/reporting_bugs.md).
- [X] Existing open issues along with etcd [frequently asked questions](https://etcd.io/docs/latest/faq) have been checked and this is not a duplicate.

### What happened?

Multiple etcd servers repeatedly panic with a message "tocommit(4432450) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?"

The first panic occurred one month after the etcd started running. At that time, there were some network issues, leading to TCP reconnections.

### What did you expect to happen?

No panic.

### How can we reproduce it (as minimally and precisely as possible)?

It's hard to reproduce the panic in real world, but we have reproduced it in a unit test.

### Anything else we need to know?

After investigation, we suspect that a malfunctioning switch or a similar issue may be dispatching incorrect Raft messages, leading to the panic. Currently, etcd does not check the `To` field in messages.

### Etcd version (please run commands below)

<details>

We ([tikv/pd](https://github.com/tikv/pd/)) use embed etcd v3.4.21 https://github.com/etcd-io/etcd/commit/85b640cee793

</details>

### Etcd configuration (command line flags or environment variables)

<details>

Default settings. 

</details>

### Etcd debug information (please run commands below, feel free to obfuscate the IP address or FQDN in the output)

<details>

```console
$ etcdctl member list -w table
# paste output here

$ etcdctl --endpoints=<member list> endpoint status -w table
# paste output here
```

</details>

### Relevant log output

```Shell
[2023/10/30 14:48:34.402 +08:00] [WARN] [stream.go:266] ["closed TCP streaming connection with remote peer"] [stream-writer-type="stream MsgApp v2"] [local-member-id=d4c1cc21d4c817ca] [remote-peer-id=18cf5532a87c6527]
[2023/10/30 14:48:34.402 +08:00] [WARN] [stream.go:277] ["established TCP streaming connection with remote peer"] [stream-writer-type="stream MsgApp v2"] [local-member-id=d4c1cc21d4c817ca] [remote-peer-id=18cf5532a87c6527]
[2023/10/30 14:48:34.409 +08:00] [INFO] [stream.go:425] ["established TCP streaming connection with remote peer"] [stream-reader-type="stream MsgApp v2"] [local-member-id=d4c1cc21d4c817ca] [remote-peer-id=18cf5532a87c6527]
[2023/10/30 14:48:34.409 +08:00] [INFO] [stream.go:425] ["established TCP streaming connection with remote peer"] [stream-reader-type="stream MsgApp v2"] [local-member-id=d4c1cc21d4c817ca] [remote-peer-id=1baed0b840546f7]
[2023/10/30 14:48:34.410 +08:00] [INFO] [stream.go:425] ["established TCP streaming connection with remote peer"] [stream-reader-type="stream Message"] [local-member-id=d4c1cc21d4c817ca] [remote-peer-id=18cf5532a87c6527]
[2023/10/30 14:48:34.410 +08:00] [INFO] [stream.go:425] ["established TCP streaming connection with remote peer"] [stream-reader-type="stream Message"] [local-member-id=d4c1cc21d4c817ca] [remote-peer-id=1baed0b840546f7]
[2023/10/30 14:48:34.419 +08:00] [PANIC] [log.go:203] ["tocommit(4432450) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?"]
...
[2023/10/30 14:48:34.419 +08:00] [PANIC] [log.go:203] ["tocommit(4432450) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?"]
...
[2023/10/30 14:48:54.770 +08:00] [PANIC] [log.go:203] ["tocommit(4432459) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?"]
...
[2023/10/30 14:49:13.070 +08:00] [PANIC] [log.go:203] ["tocommit(4432492) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?"]
...
[2023/10/30 14:49:31.681 +08:00] [PANIC] [log.go:203] ["tocommit(4432521) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?"]
...
[2023/10/30 14:49:49.816 +08:00] [PANIC] [log.go:203] ["tocommit(4432522) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?"]


panic: tocommit(4432450) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?

goroutine 249 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc00456e9c0, {0x0, 0x0, 0x0})
        /go/pkg/mod/go.uber.org/zap@v1.19.1/zapcore/entry.go:232 +0x44c
go.uber.org/zap.(*SugaredLogger).log(0xc0002bad50, 0x4, {0x2b11764?, 0xdcbc85?}, {0xc007dac240?, 0x26ef300?, 0xc0098aa300?}, {0x0, 0x0, 0x0})
        /go/pkg/mod/go.uber.org/zap@v1.19.1/sugar.go:227 +0xee
go.uber.org/zap.(*SugaredLogger).Panicf(...)
        /go/pkg/mod/go.uber.org/zap@v1.19.1/sugar.go:159
go.etcd.io/etcd/pkg/logutil.(*zapRaftLogger).Panicf(0x439fec?, {0x2b11764?, 0x2b?}, {0xc007dac240?, 0x1672334?, 0x3cd6081?})
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/pkg/logutil/zap_raft.go:101 +0x49
go.etcd.io/etcd/raft.(*raftLog).commitTo(0xc000760150, 0x43a001)
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/raft/log.go:203 +0x103
go.etcd.io/etcd/raft.(*raft).handleHeartbeat(_, {0x8, 0x18cf5532a87c6527, 0xd4c1cc21d4c817ca, 0x18, 0x0, 0x0, {0x0, 0x0, 0x0}, ...})
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/raft/raft.go:1389 +0x45
go.etcd.io/etcd/raft.stepCandidate(_, {0x8, 0x18cf5532a87c6527, 0xd4c1cc21d4c817ca, 0x18, 0x0, 0x0, {0x0, 0x0, 0x0}, ...})
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/raft/raft.go:1289 +0x585
go.etcd.io/etcd/raft.(*raft).Step(_, {0x8, 0x18cf5532a87c6527, 0xd4c1cc21d4c817ca, 0x18, 0x0, 0x0, {0x0, 0x0, 0x0}, ...})
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/raft/raft.go:990 +0x1055
go.etcd.io/etcd/raft.(*node).run(0xc000118d80)
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/raft/node.go:354 +0xa05
created by go.etcd.io/etcd/raft.RestartNode
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/raft/node.go:242 +0x24a
```

### PR DESCRIPTION
**Pull request by overvenus**
Ignore Raft messages when the `To` field mismatches the local member ID. In cases where incorrect Raft messages are dispatched, potentially due to a malfunctioning switch, this proactive check prevents panics, such as "tocommit is out of range".

Fix #16220
Fix #17081

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Review by serathius — APPROVED** [review]
High level the change looks good.

 I also did some digging ito `Process` validation.
It was introduced in https://github.com/etcd-io/etcd/pull/1343#issuecomment-60028973 with promised followups. Don't think they were realised though.

File: server/etcdserver/server_test.go [hunk 1]
```diff
@@ -889,6 +895,56 @@ func TestAddMember(t *testing.T) {
...
+		cluster:      cl,
+		reqIDGen:     idutil.NewGenerator(0, time.Time{}),
+		SyncTicker:   &time.Ticker{},
+		consistIndex: cindex.NewFakeConsistentIndex(0),
+		beHooks:      serverstorage.NewBackendHooks(lg, nil),
+	}
+	// Mock a mad switch dispatching messages to wrong node.
+	m := raftpb.Message{
+		Type:   raftpb.MsgHeartbeat,
+		To:     2, // Wrong ID.
```
[the remark this answers is not in the document]
Comment (overvenus) [reply server/etcdserver/server_test.go]: Fixed in 9f82390, actually I want the server member ID to be 1.

**Review by ahrtr — APPROVED** [review]
LGTM

We also need to backport the fix to both 3.4 and 3.5

**Comment by ahrtr** [comment]
Followup:
- backport the fix to both 3.5 and 3.4
- update changelog for both 3.5 and 3.4

@overvenus could you please raise a ticket to track the todo items?

**Comment by serathius** [comment]
cc @yichengq @jonboulle @xiang90

**Comment by jkumarpa** [comment]
@overvenus @ahrtr 
Could you please rollout these changes ? I see it's not yet backported

**Comment by ahrtr** [comment]
> @overvenus @ahrtr Could you please rollout these changes ? I see it's not yet backported

Thanks for the reminder. Contribution is welcome for https://github.com/etcd-io/etcd/issues/17081

**Comment by henrybear327** [comment]
> > @overvenus @ahrtr Could you please rollout these changes ? I see it's not yet backported
> 
> Thanks for the reminder. Contribution is welcome for #17081

Hey @ahrtr I can take this backporting task :) 

/assign

**Comment by Black-max12138** [comment]
We have a 3-node etcd cluster, and the primary node has failed. The other two nodes are unable to elect a new primary node, rendering the cluster unavailable and continuously reporting errors. 
The log error for etcd1 is
```shell
{"level":"info","ts":"2025-07-19T06:21:09.706427Z","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"9db5890b6392f376 became pre-candidate at term 2"}
{"level":"info","ts":"2025-07-19T06:21:09.706475Z","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"9db5890b6392f376 received MsgPreVoteResp from 9db5890b6392f376 at term 2"}
{"level":"info","ts":"2025-07-19T06:21:09.706549Z","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"9db5890b6392f376 [logterm: 2, index: 465746] sent MsgPreVote request to 5f23d73292959784 at term 2"}
{"level":"info","ts":"2025-07-19T06:21:09.70656Z","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"9db5890b6392f376 [logterm: 2, index: 465746] sent MsgPreVote request to 9d8e4977362dd913 at term 2"}
{"level":"warn","ts":"2025-07-19T06:21:09.706833Z","caller":"etcdserver/server.go:1015","msg":"rejected Raft message to mismatch member","local-member-id":"9db5890b6392f376","mismatch-member-id":"9d8e4977362dd913"}
{"level":"warn","ts":"2025-07-19T06:21:09.706852Z","caller":"rafthttp/http.go:145","msg":"failed to process Raft message","local-member-id":"9db5890b6392f376","error":"cannot process message to mismatch member"}
```
The log error for etcd2 is
```shell
{"level":"info","ts":"2025-07-19T07:23:03.452168Z","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"9d8e4977362dd913 is starting a new election at term 2"}
{"level":"info","ts":"2025-07-19T07:23:03.452192Z","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"9d8e4977362dd913 became pre-candidate at term 2"}
{"level":"info","ts":"2025-07-19T07:23:03.452223Z","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"9d8e4977362dd913 received MsgPreVoteResp from 9d8e4977362dd913 at term 2"}
{"level":"info","ts":"2025-07-19T07:23:03.452234Z","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"9d8e4977362dd913 [logterm: 2, index: 465746] sent MsgPreVote request to 5f23d73292959784 at term 2"}
{"level":"info","ts":"2025-07-19T07:23:03.452241Z","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"9d8e4977362dd913 [logterm: 2, index: 465746] sent MsgPreVote request to 9db5890b6392f376 at term 2"}
{"level":"warn","ts":"2025-07-19T07:23:03.452443Z","caller":"etcdserver/server.go:1015","msg":"rejected Raft message to mismatch member","local-member-id":"9d8e4977362dd913","mismatch-member-id":"9db5890b6392f376"}
{"level":"warn","ts":"2025-07-19T07:23:03.452466Z","caller":"rafthttp/http.go:145","msg":"failed to process Raft message","local-member-id":"9d8e4977362dd913","error":"cannot process message to mismatch member"}
```
The etcd cluster information is as follows:
```shell
+---------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+-----------------------+
|         ENDPOINT          |        ID        | VERSION | DB SIZE | IS LEADER | IS LEARNER | RAFT TERM | RAFT INDEX | RAFT APPLIED INDEX |        ERRORS         |
+---------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+-----------------------+
| https://172.4.63.147:2379 | 9db5890b6392f376 |  3.5.11 |  228 MB |     false |      false |         2 |     465746 |             465746 | etcdserver: no leader |
| https://172.4.63.148:2379 | 9d8e4977362dd913 |  3.5.11 |  228 MB |     false |      false |         2 |     465746 |             465746 | etcdserver: no leader |
+---------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+-----------------------+
```
@overvenus @henrybear327

**Pull request merged**